transaction/std/src/signed_contingent_input_builder.rs

```diff
@@ -0,0 +1,2272 @@
+
+            // The output should belong to the correct recipient.
+            assert!(
+                subaddress_matches_tx_out(&recipient, DEFAULT_SUBADDRESS_INDEX, &output).unwrap()
+            );
+
+            // The output should have the correct value and confirmation number
+            {
+                let public_key = RistrettoPublic::try_from(&output.public_key).unwrap();
+                assert!(confirmation.validate(&public_key, recipient.view_private_key()));
+            }
+
+            // The output's fog hint should contain the correct public key.
+            {
+                let mut output_fog_hint = FogHint::new(RistrettoPublic::from_random(&mut rng));
+                assert!(bool::from(FogHint::ct_decrypt(
+                    &ingest_private_key,
+                    &output.e_fog_hint,
+                    &mut output_fog_hint,
+                )));
+                assert_eq!(
+                    output_fog_hint.get_view_pubkey(),
+                    &CompressedRistrettoPublic::from(
+                        recipient.default_subaddress().view_public_key()
+                    )
+                );
+            }
+        }
+    }
+
+    #[test]
+    // Test a signed contingent input with two fog recipients
+    fn test_two_fogs_signed_contingent_input() {
+        let mut rng: StdRng = SeedableRng::from_seed([1u8; 32]);
+
+        for block_version in 3..=*BlockVersion::MAX {
+            let block_version = BlockVersion::try_from(block_version).unwrap();
+
+            let sender = AccountKey::new_with_fog(
+                &FromRandom::from_random(&mut rng),
+                &FromRandom::from_random(&mut rng),
+                "fog://demo.com".to_string(),
+                Default::default(),
+                vec![],
+            );
+            let recipient = AccountKey::random_with_fog(&mut rng);
+            let ingest_private_key1 = RistrettoPrivate::from_random(&mut rng);
+            let ingest_private_key2 = RistrettoPrivate::from_random(&mut rng);
+
+            let fog_resolver = MockFogResolver(btreemap! {
+                                sender
+                        .default_subaddress()
+                        .fog_report_url()
+                        .unwrap()
+                        .to_string()
+                =>
+                    FullyValidatedFogPubkey {
+                        pubkey: RistrettoPublic::from(&ingest_private_key1),
+                        pubkey_expiry: 1000,
+                    },
+                                recipient
+                        .default_subaddress()
+                        .fog_report_url()
+                        .unwrap()
+                        .to_string()
+                =>
+                    FullyValidatedFogPubkey {
+                        pubkey: RistrettoPublic::from(&ingest_private_key2),
+                        pubkey_expiry: 1500,
+                    },
+            });
```

Comment:
This pattern seems more readable with `from_iter`

```suggestion
            let fog_resolver = MockFogResolver(BTreeMap::from_iter([
                (
                    sender
                        .default_subaddress()
                        .fog_report_url()
                        .unwrap()
                        .to_string(),
                    FullyValidatedFogPubkey {
                        pubkey: RistrettoPublic::from(&ingest_private_key1),
                        pubkey_expiry: 1000,
                    },
                ),
                (
                    recipient
                        .default_subaddress()
                        .fog_report_url()
                        .unwrap()
                        .to_string(),
                    FullyValidatedFogPubkey {
                        pubkey: RistrettoPublic::from(&ingest_private_key2),
                        pubkey_expiry: 1500,
                    },
                ),
            ]));
```

Reply:
interesting -- yeah maybe there's no good reason to use `maplit`.

We are using `maplit` in many places in the test code now, maybe we can replace all uses of that as you suggest. But let's do it in a separate ticket.

Reply:
this is #1942